packages/rest/src/rest.server.ts

```diff
@@ -338,20 +343,42 @@ export class RestServer extends Context implements Server, HttpServerLike {
   }
 
   protected _setupHandlerIfNeeded() {
-    // TODO(bajtos) support hot-reloading of controllers
-    // after the app started. The idea is to rebuild the HttpHandler
-    // instance whenever a controller was added/deleted.
-    // See https://github.com/strongloop/loopback-next/issues/433
     if (this._httpHandler) return;
+    if (this._controllersAndRoutesView == null) {
+      // Watch for binding events
+      this._controllersAndRoutesView = this.createView(
+        binding =>
+          filterByKey(RestBindings.API_SPEC.key)(binding) ||
+          (filterByKey(/^(controllers|routes)\..+/)(binding) &&
+            // Exclude controller routes to avoid circular events
+            !filterByTag(RestTags.CONTROLLER_ROUTE)(binding)),
+      );
+    }
+    this._createHttpHandler();
+
+    // See https://github.com/strongloop/loopback-next/issues/433
+    this._controllersAndRoutesView.on('refresh', () => {
+      // Rebuild the HttpHandler instance whenever a controller/route was
+      // added/deleted.
+      this._createHttpHandler();
+    });
+  }
 
+  private _createHttpHandler() {
     /**
      * Check if there is custom router in the context
      */
     const router = this.getSync(RestBindings.ROUTER, {optional: true});
     const routingTable = new RoutingTable(router, this._externalRoutes);
 
     this._httpHandler = new HttpHandler(this, this.config, routingTable);
-    for (const b of this.find('controllers.*')) {
+
+    // Remove controller routes
+    for (const b of this.findByTag(RestTags.CONTROLLER_ROUTE)) {
+      this.unbind(b.key);
```

Comment:
Does this mean all routes of all controllers are unregisterd and registerd again when a new controller gets added? What happens if requests arrive in between?

Reply:
Bindings for controller based routes have to be refreshed to reflect currently registered controllers. The recycling is synchronous. No request is handled by the event loop before it finishes.

Reply:
Thanks for explanation.